File: 백준/Gold/5904. Moo 게임/Moo 게임.java
import java.util.Scanner;

public class Main {
    static int N;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N=sc.nextInt();

        int len =3;
        int prev=3;
        int k=0;

        while(len<N){
            k++;
            prev=len;
            len=2*prev+(1+2+k);
        }

        Moo(len,k);


    }
    static void Moo(int len,int k){
        int prev = (len-(1+2+k))/2;
        if(k==0){
            if(N==1){
                System.out.println("m");
                return;
            }else{
                System.out.println("o");
                return;
            }
        }
        if(N<=prev){
            Moo(prev,k-1);
        }else if(prev+1<=N&&N<prev+(3+k)){
            if(prev+1==N){
                System.out.println("m");
            }else{
                System.out.println("o");
            }
        }else{
            N-=(prev+(3+k));
            Moo(prev,k-1);
        }
    }

}

File: 백준/Gold/5904. Moo 게임/Moo 게임.cc
#include <iostream>
#include<algorithm>
using namespace std;

int N;

void MOO(int len, int n) {
	int prev = (len - (3 + n)) / 2;
	if (n == 0) {
		if (N == 1) {
			cout << "m" << '\n';
			return;
		}
		else {
			cout << "o" << '\n';
			return;
		}
	}

	if (N <= prev) {
		MOO(prev, n - 1);
	}
	else if (prev + 1 <= N && N < prev + (3 + n)) {
		if (prev + 1 == N) {
			cout << "m" << '\n';
		}
		else {
			cout << "o" << '\n';
		}
	}
	else {
		N -= (prev + 3 + n);
		MOO(prev, n - 1);
	}
}

int main(void)
{
	cin.tie(NULL)->sync_with_stdio(false);

	cin >> N;

	int len = 3;
	int prev = 3;
	int n = 0;
	while (len<N) {
		n++;
		prev = len;
		len = 2 * prev + 3 + n;
	}

	MOO(len, n);

	return 0;
}